src/feature/relay/circuitbuild_relay.c

```diff
@@ -258,13 +266,61 @@ circuit_open_connection_for_extend(const struct extend_cell_t *ec,
     return;
   }
 
+  /* Check the addresses, without logging */
+  const int ipv4_valid =
+    (circuit_extend_addr_port_helper(&ec->orport_ipv4, false, false, 0) == 0);
+  const int ipv6_valid =
+    (circuit_extend_addr_port_helper(&ec->orport_ipv6, false, false, 0) == 0);
+
+  IF_BUG_ONCE(!ipv4_valid && !ipv6_valid) {
+    /* circuit_extend_lspec_valid_helper() should have caught this */
+    circuit_mark_for_close(circ, END_CIRC_REASON_CONNECTFAILED);
+    return;
+  }
+
+  /* If we could make an IPv4 or an IPv6 connection, make an IPv6 connection
+   * at random, with probability 1 in N.
+   *   1 means "always IPv6 (and no IPv4)"
+   *   2 means "equal probability of IPv4 or IPv6"
+   *   ... (and so on) ...
+   *   (UINT_MAX - 1) means "almost always IPv4 (and almost never IPv6)"
+   * To disable IPv6, set ipv6_supported to 0.
+   */
+#define IPV6_CONNECTION_ONE_IN_N 2
+
+  const bool ipv6_supported = router_has_advertised_ipv6_orport(get_options());
```

Comment:
I think we should have a different function for this. For now it can just alias `router_has_advertised_ipv6_orport`, but someday we'll want to separate the idea of "I can connect to IPv6 addresses" and "I have an IPv6 address."

Reply:
Sounds sensible.

Let's call this one "router_can_extend_to_ipv6_orports()", because it's also possible that:
* relays or clients try connecting to IPv6 ORPorts, to see if they themselves have IPv6
* clients may be configured to connect to IPv6 ORPorts, without extending

Reply:
Used a slightly different name in 15a4180a7e041531d6923a41684e9e6ffc833760, also renamed all the mocking functions and variables.